### PR TITLE
8323809: Serial: Refactor card table verification

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple refactoring and remove much outdated comments.

Edit: The added code is mostly taken from `PSCardTable::verify_all_young_refs_imprecise`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323809](https://bugs.openjdk.org/browse/JDK-8323809): Serial: Refactor card table verification (**Enhancement** - P4)

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17447/head:pull/17447` \
`$ git checkout pull/17447`

Update a local copy of the PR: \
`$ git checkout pull/17447` \
`$ git pull https://git.openjdk.org/jdk.git pull/17447/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17447`

View PR using the GUI difftool: \
`$ git pr show -t 17447`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17447.diff">https://git.openjdk.org/jdk/pull/17447.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17447#issuecomment-1893877232)

### CODE REVIEW

File: src/hotspot/share/gc/serial/cardTableRS.cpp [hunk 1]
```diff
@@ -72,274 +72,58 @@ void CardTableRS::maintain_old_to_young_invariant(Generation* old_gen, bool is_y
...
-    assert(jp >= _begin && jp < _end,
-           "Error: jp " PTR_FORMAT " should be within "
-           "[_begin, _end) = [" PTR_FORMAT "," PTR_FORMAT ")",
-           p2i(jp), p2i(_begin), p2i(_end));
     oop obj = RawAccess<>::oop_load(p);
-    guarantee(obj == nullptr || cast_from_oop<HeapWord*>(obj) >= _boundary,
-              "pointer " PTR_FORMAT " at " PTR_FORMAT " on "
-              "clean card crosses boundary" PTR_FORMAT,
-              p2i(obj), p2i(jp), p2i(_boundary));
+    if (_young_gen->is_in_reserved(obj) &&
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/cardTableRS.cpp:82]: I would prefer, if this code also failed on values outside of the heap apart from `nullptr` for `obj`, but I understand it's not he point of this particular verification code.

File: src/hotspot/share/gc/serial/cardTableRS.cpp [hunk 2]
```diff
@@ -72,274 +72,58 @@ void CardTableRS::maintain_old_to_young_invariant(Generation* old_gen, bool is_y
...
-          if (s->block_is_obj(cur) && s->obj_is_alive(cur)) {
-            cast_to_oop(cur)->oop_iterate(&verify_blk, mr);
-          }
-        }
+    void do_object(oop obj) override {
+      CheckForUnmarkedOops object_check(_young_gen, _card_table);
+      obj->oop_iterate(&object_check);
+      if (object_check.has_unmarked_oop()) {
+        guarantee(_card_table->is_dirty_for_addr(obj), "Found unmarked old-to-young pointer");
       }
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/cardTableRS.cpp:122]:
```suggestion
      guarantee(!object_check.has_unmarked_oop(), "Found unmarked old-to-young pointer");
```

This looks much more concise, there is no need to repeat the check imo. Please also add some information about the failing address into the message.

File: src/hotspot/share/gc/serial/cardTableRS.cpp [hunk 3]
```diff
@@ -72,274 +72,58 @@ void CardTableRS::maintain_old_to_young_invariant(Generation* old_gen, bool is_y
...
-           "Error: jp " PTR_FORMAT " should be within "
-           "[_begin, _end) = [" PTR_FORMAT "," PTR_FORMAT ")",
-           p2i(jp), p2i(_begin), p2i(_end));
     oop obj = RawAccess<>::oop_load(p);
-    guarantee(obj == nullptr || cast_from_oop<HeapWord*>(obj) >= _boundary,
-              "pointer " PTR_FORMAT " at " PTR_FORMAT " on "
-              "clean card crosses boundary" PTR_FORMAT,
-              p2i(obj), p2i(jp), p2i(_boundary));
+    if (_young_gen->is_in_reserved(obj) &&
+        !_card_table->is_dirty_for_addr(p)) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/cardTableRS.cpp:83]: Is this correct for imprecise marks at e.g. before-gc verification?

In this case `p` might be on a different card than the mark. The original code seems to handle this case (although clumsily afaict). Or am I missing something?

How did you test this change?

File: src/hotspot/share/gc/serial/cardTableRS.cpp [hunk 4]
```diff
@@ -72,274 +72,58 @@ void CardTableRS::maintain_old_to_young_invariant(Generation* old_gen, bool is_y
...
+  CheckForUnmarkedOops(DefNewGeneration* young_gen, CardTableRS* card_table) :
+    _young_gen(young_gen),
+    _card_table(card_table),
+    _unmarked_addr(nullptr) {}
 
-  virtual void do_oop(oop* p)       { VerifyCleanCardClosure::do_oop_work(p); }
-  virtual void do_oop(narrowOop* p) { VerifyCleanCardClosure::do_oop_work(p); }
-};
+  void do_oop(oop* p)       override { CheckForUnmarkedOops::do_oop_work(p); }
+  void do_oop(narrowOop* p) override { CheckForUnmarkedOops::do_oop_work(p); }
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/cardTableRS.cpp:98]:
```suggestion
  void do_oop(oop* p)       override { do_oop_work(p); }
  void do_oop(narrowOop* p) override { do_oop_work(p); }
```

Should work...

**Review by tschatzl — CHANGES_REQUESTED** [review]
I did not really try to understand the original code in complete detail, but focused myselves on the new code where the point of the verification is that for every old-to-young reference there must be a dirty card mark.
In particular it is unclear to me how the code handles imprecise marks.

File: src/hotspot/share/gc/serial/cardTableRS.cpp (hunk 2 above)
Comment (albertnetymk) [reply src/hotspot/share/gc/serial/cardTableRS.cpp:123]: The `if` is required to account for imprecise-mark -- obj has no unmarked oops or obj-start is dirty.

**Review by tschatzl — APPROVED** [review]
Thanks for the updates. Lgtm.

**Review by kstefanj — APPROVED** [review]
Looks good.

**Comment by albertnetymk** [comment]
Thanks for review.

/integrate

**Pull request closed without merging**